### PR TITLE
Huge refactor of buildscript

### PR DESCRIPTION
**Pull request by solonovamax**
- Store versions in version catalogs
- Move any duplicate logic to convention plugins
- Refactor modules
  - Refactor main project -> api subproject
  - Refactor compose subproject -> gui subproject
- Add axion-release plugin for versioning
- Add typesafe project accessors for depending on subprojects
- Add dokka for documentation generation
- Clean up & refactor publishing

Tell me if there's anything you'd like changed

### CODE REVIEW

File: buildSrc/src/main/kotlin/jneedle.publishing.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                    }
+                }
+                developers {
+                    developer {
+                        id.set("kosmx")
+                        name.set("KosmX")
+                        email.set("kosmx.mc@gmail.com")
+                    }
+
+                    developer {
```
Comment (KosmX) [inline buildSrc/src/main/kotlin/jneedle.publishing.gradle.kts:47]: Ok, this should be moved to a contributors list file, but that can wait

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+kotlinx-serialization-core = { group = "org.jetbrains.kotlinx", name = "kotlinx-serialization-core", version.ref = "kotlinx-serialization" }
+kotlinx-serialization-json = { group = "org.jetbrains.kotlinx", name = "kotlinx-serialization-json", version.ref = "kotlinx-serialization" }
+kotlinx-serialization-cbor = { group = "org.jetbrains.kotlinx", name = "kotlinx-serialization-cbor", version.ref = "kotlinx-serialization" }
+kotlinx-serialization-hocon = { group = "org.jetbrains.kotlinx", name = "kotlinx-serialization-hocon", version.ref = "kotlinx-serialization" }
+kotlinx-serialization-protobuf = { group = "org.jetbrains.kotlinx", name = "kotlinx-serialization-protobuf", version.ref = "kotlinx-serialization" }
+
+# TODO: decide whether to use kotlinx.serialization for config or konf (or maybe smth else?)
+# ktoml
+## TOML serialization for kotlinx.serialization
+ktoml-core = { group = "com.akuleshov7", name = "ktoml-core", version.ref = "ktoml" }
```
Comment (KosmX) [inline gradle/libs.versions.toml:49]: There is a lot of stuff here apparently unused...

**Comment by KosmX** [comment]
Actually, the gradle buildSrc is good, it will keep gradle readable.  
I don't see how is changing imports a part of gradle buildscript rewrite, but I'll refactor the whole project anyway, so it doesn't matter.  

Adding yourself to contributors is okay (really, I would do the same thing), but including/promoting your maven repo is unnecessary, it's just one more point of possible failure.

File: buildSrc/src/main/kotlin/jneedle.publishing.gradle.kts (hunk 1 above)
Comment (solonovamax) [reply buildSrc/src/main/kotlin/jneedle.publishing.gradle.kts:47]: tbh it kinda *is* in its own file already lol

File: buildSrc/src/main/kotlin/Utilities.kt
[the remark this answers is not in the document]
Comment (solonovamax) [reply buildSrc/src/main/kotlin/Utilities.kt:1]: funny gradle magic go brrrr

File: buildSrc/src/main/kotlin/jneedle.publishing.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        maven {
+            name = "Sonatype"
+
+            val releasesUrl = uri("https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/") // releases repo
+            val snapshotUrl = uri("https://s01.oss.sonatype.org/content/repositories/snapshots/") // snapshot repo
+            url = if (isSnapshot) snapshotUrl else releasesUrl
+
+            credentials(PasswordCredentials::class)
+        }
+        maven {
```
[the remark this answers is not in the document]
Comment (solonovamax) [reply buildSrc/src/main/kotlin/jneedle.publishing.gradle.kts:76]: it's not particularly "random repos"
that's my own repo that I use for projects

but yeah, I forgor to remove it lol

File: buildSrc/src/main/kotlin/jneedle.repositories.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+repositories {
+    maven("https://maven.solo-studios.ca/releases/") {
```
[the remark this answers is not in the document]
Comment (solonovamax) [reply buildSrc/src/main/kotlin/jneedle.repositories.gradle.kts:2]: I mean, if we did end up using slf4k, it's in that repo, but it's also in maven central


but, fair

File: gradle/libs.versions.toml (hunk 2 above)
Comment (solonovamax) [reply gradle/libs.versions.toml:49]: yeah

I felt it might be better to put in a bunch of stuff to "look at using" and then slim it down afterwards

**Comment by solonovamax** [comment]
> but including/promoting your maven repo is unnecessary, it's just one more point of possible failure.

fair
I'll remove it when I'm on my pc next

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+##  A modern I/O library for Android, Java, and Kotlin Multiplatform.
+okio = { group = "com.squareup.okio", name = "okio", version = "3.3.0" }
+
+
+# SLF4J
+## Simple Logging Facade for Java
+slf4j = { group = "org.slf4j", name = "slf4j-api", version = "2.0.7" }
+# SLF4K
+## SLF4K is a set of kotlin extensions for SLF4J
+slf4k = { group = "ca.solo-studios", name = "slf4k", version = "0.5.3" }
```
[the remark this answers is not in the document]
Comment (solonovamax) [reply gradle/libs.versions.toml:121]: it's my own little wrapper for slf4j

https://github.com/solo-studios/slf4k

**Comment by solonovamax** [comment]
also btw if you feel like it, you should have write access to this branch so you should be able to push changes to it if you wanted to, but I'll do it in like ~5-ish hours when I can

**Comment by solonovamax** [comment]
oopsies, CI failed lol

I know what the issue is (it's the stupid gradle toolchains thing)

**Comment by KosmX** [comment]
Gradle caching is already in github actions, just PRs can't use project cache for security reasons...

**Review by KosmX — APPROVED** [review]

**Pull request merged**